### PR TITLE
Canvas exempting and seeding changes

### PR DESCRIPTION
**Pull request by jonathanKingston**
**Reviewer:**

## Automated tests:
- [ ] Unit tests
- [ ] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

File: shared/js/content-scope/fingerprinting-canvas-protection.js [hunk 1]
```diff
@@ -6,25 +6,69 @@ export function init (args) {
...
+                clearCache(thisArg.canvas)
+                return DDGReflect.apply(target, thisArg, args)
+            }
+        })
+        debuggingProxy.overload()
+    }
+    // }
+
+    // Include all these: https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D
+    // Or overload the parent object and make an allow list of the put and draw calls only?
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply shared/js/content-scope/fingerprinting-canvas-protection.js:30]: The allowlist is certainly better, however that means changing DDG proxy to work with `get` also and then change other implementations of the `apply` in this file. I think we could do this improvement as a follow up.

File: shared/js/content-scope/utils.js [hunk 2]
```diff
@@ -144,10 +152,11 @@ export class DDGProxy {
         this.objectScope = objectScope
         this.property = property
         this.featureName = featureName
+        this.camelFeatureName = camelcase(this.featureName)
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply shared/js/content-scope/utils.js:155]: The `shouldExemptMethod` was being called with the dasherised form so not currently working.

File: scripts/inject.mjs [hunk 3]
```diff
@@ -28,7 +30,9 @@ async function generateContentScope() {
         dir: 'build',
         format: 'iife',
         inlineDynamicImports: true,
-        name: 'protections'
+        name: 'protections',
+        // This if for seedrandom causing build issues
+        globals: { crypto: 'undefined' }
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply scripts/inject.mjs:35]: They use this for improved seeding but it's not required, it's just a dependency that breaks the build.

File: shared/js/content-scope/fingerprinting-canvas-protection.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (englehardt) [reply shared/js/content-scope/fingerprinting-canvas-protection.js:30]: Sounds good! Let's at least add the functions I mentioned as part of this PR.

File: integration-test/content-scripts/test-canvas.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            'good.third-party.site',
+            'broken.third-party.site'
+        ]
+        const hostnameResults = {}
+        for (const hostname of hostnames) {
+            const page = await browser.newPage()
+            await page.goto(`https://${hostname}/features/canvas-draw.html`, { waitUntil: 'networkidle0' })
+            // Wait for injection; will be resolved with MV3 changes
+            await page.waitForFunction(
+                () => navigator.globalPrivacyControl,
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply integration-test/content-scripts/test-canvas.js:40]: We check https elsewhere for the background too. We don't use it elsewhere and admittedly not ideal, however more preferred over setTImeout.

**Review by englehardt — APPROVED** [review]
This LGTM. Thanks for all of the updates.

I also tested locally briefly, checking to make sure I can run the canvas tests in the privacy test page repo as expected and checked against https://browserleaks.com/canvas.

**Pull request merged**